lib/veri.ml

```diff
@@ -148,41 +152,33 @@ class ['a] t arch dis =
     method private update_event ev =
       SM.update (fun c -> c#register_event ev)
 
-    (** [resolve_var var] - returns a result, bound with [var].
-        Sequence of searches is the following:
-        1) among read events that occured at current step in the same context,
-           with the same variable;
-        2) among read events that occures at current step, in other context,
-           with the same variable;
-        3) in current context, for the same variable *)
-    method private resolve_var : var -> 'a r = fun var ->
+  (** [find_var_data var] - returns a result, bound with [var].
+      Sequence of searches is the following:
+      1) among write events that occured at current step in the same context,
+         with the same variable; we don't need to emit event in this case
+      2) among read events that occured at current step in the same context,
+         with the same variable; will emit register read in this case
+      3) among read events that occures at current step, in other context,
+         with the same variable; will emit register read in this case *)
+    method private find_var_data var =
       SM.get () >>= fun ctxt ->
-      match find_reg_read (self_events ctxt) (same_var var) with
-      | Some mv -> self#eval_exp (Bil.int (Move.data mv))
-      | None ->
-        match find_reg_read (other_events ctxt) (same_var var) with
-        | Some mv -> self#eval_exp (Bil.int (Move.data mv))
-        | None -> super#lookup var
-
-    (** [lookup var] - returns a result, bound with variable.
-        Search starts from self events, if it was write access to given
-        variable at current step. And if it was, then result of write
-        access returned and no read event is emitted.
-        Otherwise searching continues as written above for [resolve_var],
-        with emitting register_read event. *)
+      let searches = [
+        (fun () -> find_reg_write (self_events ctxt)  (same_var var)), false;
+        (fun () -> find_reg_read  (self_events ctxt)  (same_var var)), true;
+        (fun () -> find_reg_read  (other_events ctxt) (same_var var)), true;
+      ] in
+      SM.return (find_move_data searches)
```

Comment:
It looks like that there is a better choice of abstraction here, as I see lot's of code repetition in the `searches` definition.